pkg/controller/seed-controller-manager/addon/addon_controller.go

```diff
@@ -564,22 +592,36 @@ func (r *Reconciler) ensureIsInstalled(ctx context.Context, log *zap.SugaredLogg
 		return fmt.Errorf("failed to create command: %w", err)
 	}
 
-	if addon.Name == "csi" &&
-		cluster.Spec.Cloud.Hetzner != nil &&
-		cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
-		cluster.Annotations[migratedHetznerCSIAddon] != "yes" {
-		// Between v2.22 and v2.23, there was a change to hetzner CSI driver immutable field fsGroupPolicy
-		// as a result, the CSDriver resource has to be redeployed
-		// https://github.com/kubermatic/kubermatic/issues/12429
-		if err := r.migrateHetznerCSIDriver(ctx, log, cluster); err != nil {
-			return fmt.Errorf("failed to migrate CSI Driver: %w", err)
-		}
-		if cluster.Annotations == nil {
-			cluster.Annotations = make(map[string]string)
-		}
-		cluster.Annotations[migratedHetznerCSIAddon] = "yes"
-		if err := r.Update(ctx, cluster); err != nil {
-			log.Errorf("failed to set %q cluster annotation: %w", migratedHetznerCSIAddon, err)
+	if addon.Name == "csi" {
+		if cluster.Spec.Cloud.Hetzner != nil &&
+			cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+			cluster.Annotations[migratedHetznerCSIAddon] != yes {
+			// Between v2.22 and v2.23, there was a change to hetzner CSI driver immutable field fsGroupPolicy
+			// as a result, the CSDriver resource has to be redeployed
+			// https://github.com/kubermatic/kubermatic/issues/12429
+			if err := r.migrateHetznerCSIDriver(ctx, log, cluster); err != nil {
+				return fmt.Errorf("failed to migrate CSI Driver: %w", err)
+			}
+			if cluster.Annotations == nil {
+				cluster.Annotations = make(map[string]string)
+			}
+			cluster.Annotations[migratedHetznerCSIAddon] = yes
+			if err := r.Update(ctx, cluster); err != nil {
+				log.Errorf("failed to set %q cluster annotation: %w", migratedHetznerCSIAddon, err)
+			}
+		} else if cluster.Spec.Cloud.VSphere != nil &&
+			cluster.Spec.Features[kubermaticv1.ClusterFeatureExternalCloudProvider] &&
+			cluster.Annotations[migratedVsphereCSIAddon] != yes {
+			if err := r.migrateVsphereCSIDriver(ctx, log, cluster); err != nil {
+				return fmt.Errorf("failed to migrate CSI Driver: %w", err)
+			}
+			if cluster.Annotations == nil {
+				cluster.Annotations = make(map[string]string)
+			}
+			cluster.Annotations[migratedVsphereCSIAddon] = yes
+			if err := r.Update(ctx, cluster); err != nil {
+				log.Errorf("failed to set %q cluster annotation: %w", migratedVsphereCSIAddon, err)
```

Comment:
Shouldn't this be `log.Errorw("...", zap.Error(err))`?

Reply:
Same with Hetzner, so I assume you copy-pasted. 😁

Reply:
damn, yes.